### PR TITLE
Email selected users when reimbursement requests are entered

### LINKED ISSUE #4476: Email selected users when reimbursement requests are entered
## What type(s) of user does this feature affect?
 - supervisors
 - admins

## Description
See: https://github.com/rubyforgood/casa/issues/4474
For each new reimbursement request submitted, send an email to admins/supervisors who have reimbursement notifications enabled.

Email Subject: " New reimbursement request from [volunteer name]"
Email text: " [Volunteer name] has submitted a reimbursement request, please follow up on the reimbursements page
- Include a link to the reimbursements page

## Screenshots of current behavior, if any
image

## QA Login Details
[Link to QA site](https://casa-qa.herokuapp.com/)  

Login Emails: 
- volunteer1@example.com  view site as a volunteer
- supervisor1@example.com view site as a supervisor
- casa_admin1@example.com view site as an admin

password for all users: 12345678  

### Questions? Join Slack!

We highly recommend that you join us in slack https://rubyforgood.herokuapp.com/ #casa channel to ask questions quickly and hear about office hours (currently Tuesday 6-8pm Pacific), stakeholder news, and upcoming new issues.

### PR DESCRIPTION
**Pull request by sarvaiyanidhi**
Resolves #4476 

### What changed, and why?
This PR implements feature of sending email on creation of new reimbursement request to Supervisor if reimbursement notifications enabled

### How will this affect user permissions?
- Volunteer permissions: Not effected
- Supervisor permissions: Not effected
- Admin permissions: Not effected

### How is this tested? (please write tests!) 💖💪
Have added required tests for email feature.

### Screenshots please :)

reimbursement_request_email

### CODE REVIEW

File: app/controllers/case_contacts_controller.rb [hunk 1]
```diff
@@ -181,6 +185,12 @@ def create_case_contact_for_every_selected_casa_case(selected_cases)
     end
   end
 
+  def send_reimbursement_email(case_contact)
+    if case_contact.want_driving_reimbursement_changed? && case_contact.want_driving_reimbursement? && !current_user.supervisor.blank?
+      SupervisorMailer.reimbursement_request_email(current_user, current_user.supervisor).deliver
```
Comment (littleforest) [inline app/controllers/case_contacts_controller.rb:190]:
```suggestion
      SupervisorMailer.reimbursement_request_email(current_user, current_user.supervisor).deliver_later
```

File: app/views/supervisor_mailer/reimbursement_request_email.html.erb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+<table width="100%" cellpadding="0" cellspacing="0" style="font-family: Helvetica,Arial,sans-serif; box-sizing: border-box; font-size: 14px; margin: 0;">
+  <tr style="font-family: Helvetica,Arial,sans-serif; box-sizing: border-box; font-size: 14px; margin: 0;">
+    <td class="content-block" style="font-family: Helvetica,Arial,sans-serif; box-sizing: border-box; font-size: 14px; vertical-align: top; margin: 0; padding: 0 0 20px;" valign="top">
+      Hello <%= @supervisor.display_name %>,
+    </td>
+  </tr>
+  <tr style="font-family: Helvetica,Arial,sans-serif; box-sizing: border-box; font-size: 14px; margin: 0;">
+    <td class="content-block" style="font-family: Helvetica,Arial,sans-serif; box-sizing: border-box; font-size: 14px; vertical-align: top; margin: 0; padding: 0 0 20px;" valign="top">
+      <%= @volunteer.display_name %> has submitted a reimbursement request, please follow up on the reimbursements page
+      <a href="<%= "#{reimbursements_url}" %>" target="_blank">using this link</a>.
```
[the remark this answers is not in the document]
Comment (littleforest) [reply app/views/supervisor_mailer/reimbursement_request_email.html.erb:12]: Best to use the `link_to` helper instead:

```
<%= link_to "using this link", reimbursements_url %>
```
I would just leave off the `target: "_blank"` as mail programs generally open links in a new window anyway.

File: spec/mailers/supervisor_mailer_spec.rb [hunk 3]
```diff
@@ -116,4 +116,16 @@
...
+
+  describe ".reimbursement_request_email" do
+    let(:supervisor) { create(:supervisor, receive_reimbursement_email: true) }
+    let(:volunteer) { create(:volunteer) }
+    let(:casa_organization) { volunteer.casa_org }
+
+    let(:mail) { SupervisorMailer.reimbursement_request_email(volunteer, supervisor) }
+
+    it "sends email reminder" do
+      expect(mail.body.encoded).to match("#{volunteer.display_name} has submitted a reimbursement request")
```
Comment (littleforest) [inline spec/mailers/supervisor_mailer_spec.rb:128]: Could you also check that the `to:` and the `subject:` are correct?

File: app/controllers/case_contacts_controller.rb (hunk 1 above)
Comment (littleforest) [reply app/controllers/case_contacts_controller.rb:190]: Could you also add/modify `spec/requests/case_contacts_spec.rb` and add an expectation that `SupervisorMailer.reimbursement_request_email` is called when those conditions are met, and another expection that it is not called when the conditions are not met?

**Review by littleforest — COMMENTED** [review]
Thanks for working on this!

File: app/controllers/case_contacts_controller.rb (hunk 1 above)
Comment (sarvaiyanidhi) [reply app/controllers/case_contacts_controller.rb]: Thank you for the feedback. Have done the required changes and pushed the updated code.

File: spec/mailers/supervisor_mailer_spec.rb (hunk 3 above)
Comment (sarvaiyanidhi) [reply spec/mailers/supervisor_mailer_spec.rb:130]: Have done the changes to check to and subject as well.

File: spec/mailers/previews/supervisor_mailer_preview_spec.rb [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 RSpec.describe SupervisorMailerPreview do
   let!(:supervisor) { create(:supervisor) }
+  let!(:volunteer) { create(:volunteer) }
```
Comment (schoork) [inline spec/mailers/previews/supervisor_mailer_preview_spec.rb:6]: Can you make it explicit that the volunteer's supervisor is the supervisor created above? Might get flaky in the future if we don't.

Comment (sarvaiyanidhi) [reply spec/mailers/previews/supervisor_mailer_preview_spec.rb]: Thank you @schoork for the feedback. I have addressed the reported point and pushed the fix. Docker check is failing due to other issue, can you trigger it again and review the changes again to merge it.

File: app/mailers/supervisor_mailer.rb [hunk 5]
```diff
@@ -19,4 +19,13 @@ def weekly_digest(supervisor)
...
     )
   end
+
+  def reimbursement_request_email(volunteer, supervisor)
+    @volunteer = volunteer
+    @casa_organization = volunteer.casa_org
+    @supervisor = supervisor
+    if supervisor.receive_reimbursement_email
+      mail(to: @supervisor.email, subject: "New reimbursement request from #{@volunteer.display_name}")
+    end
```
Comment (javiguerra777) [inline app/mailers/supervisor_mailer.rb:29]: Good use of an if conditional to make sure that receieve_reimbursement_email is part of the supervisor parameter before sending mail to the supervisor email

**Review by schoork — APPROVED** [review]
Great work on this!

File: app/controllers/case_contacts_controller.rb [hunk 6]
```diff
@@ -181,6 +185,12 @@ def create_case_contact_for_every_selected_casa_case(selected_cases)
     end
   end
 
+  def send_reimbursement_email(case_contact)
+    if case_contact.want_driving_reimbursement_changed? && case_contact.want_driving_reimbursement? && !current_user.supervisor.blank?
```
[the remark this answers is not in the document]
Comment (sarvaiyanidhi) [reply app/controllers/case_contacts_controller.rb]: Thank you for the great feedback and changes done in this PR. I have gone through the commit and understood the fixes done in the code.

**Pull request merged**